plugins/flatpak/gs-flatpak-transaction.c

```diff
@@ -62,6 +62,41 @@ gs_flatpak_transaction_set_no_deploy (FlatpakTransaction *transaction, gboolean
 	g_object_notify (G_OBJECT (self), "no-deploy");
 }
 
+/* Checks if a ref is a related ref to one of the installed ref.
+ * If yes, return the GsApp corresponding to the installed ref,
+ * NULL otherwise.
+ */
+static GsApp *
+get_installed_main_app_of_related_ref (FlatpakTransaction          *transaction,
+                                       FlatpakTransactionOperation *operation)
+{
+	FlatpakInstallation *installation = flatpak_transaction_get_installation (transaction);
+	const gchar *remote = flatpak_transaction_operation_get_remote (operation);
+	const gchar *op_ref = flatpak_transaction_operation_get_ref (operation);
+	GsFlatpakTransaction *self = GS_FLATPAK_TRANSACTION (transaction);
+	GList *keys;
+
+	if (g_str_has_prefix (op_ref, "app/"))
+		return NULL;
+
+	keys = g_hash_table_get_keys (self->refhash);
+	for (GList *l = keys; l != NULL; l = l->next) {
+		g_autoptr(GPtrArray) related_refs = NULL;
+		related_refs = flatpak_installation_list_installed_related_refs_sync (installation, remote,
+										      l->data, NULL, NULL);
+		if (related_refs == NULL)
+			continue;
+
+		for (guint i = 0; i < related_refs->len; i++) {
+			g_autofree gchar *rref = flatpak_ref_format_ref (g_ptr_array_index (related_refs, i));
+			if (g_strcmp0 (rref, op_ref) == 0) {
+				return g_hash_table_lookup (self->refhash, l->data);
```

Comment:
Also I believe you need to ref the result here as you are using g_autoptr(GsApp) below?

Reply:
It's already ref-ed when it is inserted in the hashtable

Reply:
Yes, but it is freed when the hashtable is unrefed on finalize - so you will end up with a double free here. You need to either ref it here or dont use g_autoptr for main_app in `gs_flatpak_transaction_get_app_by_ref()`.

Reply:
Yeah, I missed that _finalize bit it seems. Updated